In [1]:
import pandas as pd

Loading Dataset

In [2]:
df = pd.read_excel('/content/synthetic_shipping_delivery_1000.xlsx')
df

,delivery_id,delivery_time_start,delivery_time_finish,delivery_time_estimation_in_hours,delivery_time_result_in_hours,delivery_status,price,weight,product_type,place_start,place_end,rating,delivery_type,distance,delivery_way,refund,carrier_id,package_volume_m3
0,DEL00001,2026-03-16 14:00,2026-03-19 02:00,60,60,Damage,368.12,216.87,Furniture,Hub Beta,Hub Beta,4,Same Day,145.6,Sea,No,CAR008,1.554
1,DEL00002,2026-04-15 05:00,2026-04-16 02:00,20,21,Damage,190.07,204.96,Clothing,West Center,Central Hub,5,Same Day,2703.7,Mixed Way,Yes,CAR028,3.246
2,DEL00003,2026-03-05 05:00,2026-03-05 21:00,2,16,Damage,80.62,425.78,Documents,North Depot,Hub Gamma,1,Express,2910.5,Mixed Way,Yes,CAR020,2.774
3,DEL00004,2026-01-27 19:00,2026-01-29 18:00,33,47,Damage,347.36,169.49,Medicine,East Center,Central Hub,4,Standard,2441.8,Air,No,CAR009,1.207
4,DEL00005,2026-08-12 16:00,2026-08-17 10:00,104,114,Damage,80.42,338.05,Medicine,Hub Gamma,Hub Alpha,3,Standard,2154.0,Sea,No,CAR017,3.110
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,DEL00996,2026-01-02 06:00,2026-01-05 03:00,56,69,Complete,127.73,310.34,Groceries,Hub Gamma,East Center,5,Express,147.0,Mixed Way,No,CAR022,4.031
996,DEL00997,15/03/2026 09:00,2026-06-13 23:00,47,51,Damage,484.83,116.00,Electronics,North Depot,Hub Alpha,5,Standard,265.3,Road,No,CAR010,1.906
997,DEL00998,2026-04-06 21:00,2026-04-10 02:00,66,77,Complete,93.55,455.16,Furniture,Central Hub,South Depot,2,Express,998.7,Sea,Yes,CAR012,1.894
998,DEL00999,2026-06-11 13:00,2026-06-11 22:00,12,9,Damage,297.10,428.16,Documents,East Center,Hub Beta,2,Economy,351.2,Mixed Way,Yes,CAR005,4.526


Understanding the Dataset Shape and Data

Based on the code result, it is expected to contain **1,000 records**. However, there are **10 missing values** in the **Price** column and **10 missing values** in the **Weight** column.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 18 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   delivery_id                        1000 non-null   object 
 1   delivery_time_start                1000 non-null   object 
 2   delivery_time_finish               1000 non-null   object 
 3   delivery_time_estimation_in_hours  1000 non-null   int64  
 4   delivery_time_result_in_hours      1000 non-null   int64  
 5   delivery_status                    1000 non-null   object 
 6   price                              1000 non-null   float64
 7   weight                             990 non-null    float64
 8   product_type                       990 non-null    object 
 9   place_start                        1000 non-null   object 
 10  place_end                          1000 non-null   object 
 11  rating                             1000 non-null   int64 

`df["weight"].fillna(df['weight'].median(), inplace = True)` is a code to fill missing value on weight column with median value from weight column

In [4]:
df["weight"].fillna(df['weight'].median(), inplace = True)

/tmp/ipykernel_6046/3487639992.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["weight"].fillna(df['weight'].median(), inplace = True)


`df["product_type"].fillna(df['product_type'].mode()[0], inplace = True)` is a code to fill missing value on product_type column with mode value from product_type column

In [5]:
df["product_type"].fillna(df['product_type'].mode()[0], inplace = True)

/tmp/ipykernel_6046/773397762.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["product_type"].fillna(df['product_type'].mode()[0], inplace = True)


The `df.describe()` function helps users understand the numerical columns in the dataset by providing several descriptive statistics. Based on the results, we noticed that there are incorrect values in the Price column.

In [6]:
df.describe()

,delivery_time_estimation_in_hours,delivery_time_result_in_hours,price,weight,rating,distance,package_volume_m3
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000
mean,62.084000,67.131000,247.540290,243.869330,3.023000,1526.328000,2.459602
std,33.692197,34.189922,144.708428,140.332314,1.422144,864.947138,1.412577
min,2.000000,0.000000,-20.000000,0.320000,1.000000,3.600000,0.011000
25%,34.000000,39.000000,120.810000,123.822500,2.000000,802.150000,1.222500
50%,62.000000,66.000000,242.720000,238.855000,3.000000,1567.150000,2.451500
75%,90.000000,96.000000,373.882500,360.470000,4.000000,2279.000000,3.605750
max,120.000000,134.000000,499.700000,498.890000,5.000000,2992.700000,4.991000


`df['price'].replace(to_replace=-20.00,value=20.00,inplace=True)` is a code that replace -20 incorrect value on price column into 3

In [7]:
df['price'].replace(to_replace=-20.00,value=20.00,inplace=True)

/tmp/ipykernel_6046/3136492734.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['price'].replace(to_replace=-20.00,value=20.00,inplace=True)


Checking categorical value (delivery status, product_type, delivery_type, delivery_way)

In [8]:
df['delivery_status'].unique()

array(['Damage', 'Complete', 'Lost'], dtype=object)

In [9]:
df['product_type'].unique()

array(['Furniture', 'Clothing', 'Documents', 'Medicine', 'Groceries',
       'Electronics', 'Machinery', 'Books'], dtype=object)

`df["product_type"].fillna(df['product_type'].mode()[0], inplace = True)` is a code to fill missing value on product_type column with mode value from product_type column

In [10]:
df["product_type"].fillna(df['product_type'].mode()[0], inplace = True)

In [11]:
df['delivery_type'].unique()

array(['Same Day', 'Express', 'Standard', 'Economy'], dtype=object)

In [12]:
df['delivery_way'].unique()

array(['Sea', 'Mixed Way', 'Air', 'Road', 'Rail'], dtype=object)

Adding one column named order_difference, which will contain the difference between actual and estimation delivery time  

In [13]:
df['order_difference'] = df['delivery_time_result_in_hours'] - df['delivery_time_estimation_in_hours']

#EDA

Preparing for exploratory data analysis, classifying every delivery to their class (on time, delay, early)

In [14]:
df_complete = df[df['order_difference'] == 0]
df_complete["order_result"] = "on time"

/tmp/ipykernel_6046/1431261861.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_complete["order_result"] = "on time"


In [15]:
df_delay = df[df['order_difference'] > 0]
df_delay["order_result"] = "delay"

/tmp/ipykernel_6046/3002441686.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_delay["order_result"] = "delay"


In [16]:
df_early = df[df['order_difference'] < 0]
df_early["order_result"] = "early"

/tmp/ipykernel_6046/2738192414.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_early["order_result"] = "early"


In [17]:
df=pd.concat([df_complete, df_delay, df_early], axis=0)

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 0 to 999
Data columns (total 20 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   delivery_id                        1000 non-null   object 
 1   delivery_time_start                1000 non-null   object 
 2   delivery_time_finish               1000 non-null   object 
 3   delivery_time_estimation_in_hours  1000 non-null   int64  
 4   delivery_time_result_in_hours      1000 non-null   int64  
 5   delivery_status                    1000 non-null   object 
 6   price                              1000 non-null   float64
 7   weight                             1000 non-null   float64
 8   product_type                       1000 non-null   object 
 9   place_start                        1000 non-null   object 
 10  place_end                          1000 non-null   object 
 11  rating                             1000 non-null   int64  
 12

In [19]:
df.groupby(['delivery_status']).agg(
    number_of_delivery = ('delivery_id','count')
)

,number_of_delivery
delivery_status,
Complete,496
Damage,494
Lost,10


According to the data, there are 496 complete delivery, 494 damage delivery, 10 lost delivery

In [20]:
df.groupby(['delivery_status','delivery_type']).agg(
    number_of_delivery = ('delivery_id','count')
)

number_of_delivery
delivery_status delivery_type                    
Complete        Economy                       129
                Express                       131
                Same Day                      112
                Standard                      124
Damage          Economy                       125
                Express                       122
                Same Day                      128
                Standard                      119
Lost            Economy                         1
                Express                         4
                Same Day                        3
                Standard                        2

According to the data, all delivery types have a similar number of damaged deliveries (around 119–128) and lost deliveries (ranging from 1–4 deliveries).

In [21]:
df.groupby(['delivery_status','delivery_way']).agg(
    number_of_delivery = ('delivery_id','count')
)

number_of_delivery
delivery_status delivery_way                    
Complete        Air                          119
                Mixed Way                    134
                Rail                           3
                Road                         120
                Sea                          120
Damage          Air                          129
                Mixed Way                    119
                Rail                           7
                Road                         125
                Sea                          114
Lost            Air                            2
                Mixed Way                      6
                Road                           1
                Sea                            1

According to the data, **Rail** has the smallest number of completed and damaged deliveries compared to the other delivery methods. For damaged deliveries, **Air, Mixed Way, Road, and Sea** have a similar number of deliveries. For lost deliveries, **Mixed Way** has the highest number.

In [22]:
df.groupby(['delivery_status','product_type']).agg(
    number_of_delivery = ('delivery_id','count')
)

number_of_delivery
delivery_status product_type                    
Complete        Books                         59
                Clothing                      52
                Documents                     71
                Electronics                   58
                Furniture                     64
                Groceries                     71
                Machinery                     59
                Medicine                      62
Damage          Books                         64
                Clothing                      60
                Documents                     55
                Electronics                   63
                Furniture                     63
                Groceries                     68
                Machinery                     63
                Medicine                      58
Lost            Books                          1
                Clothing                       1
                Documents                      2
                Electronics                    1
                Furniture                      1
                Groceries                      2
                Machinery                      2

According to the data, all product types have a similar number of damaged and lost deliveries

In [36]:
df.groupby(['product_type']).agg(
    number_of_delivery = ('delivery_id','count')
).sort_values(by="number_of_delivery",ascending=False)

,number_of_delivery
product_type,
Groceries,141
Documents,128
Furniture,128
Books,124
Machinery,124
Electronics,122
Medicine,120
Clothing,113


According to the data, Groceries is the most commonly delivered product type, while Clothing is the least commonly delivered product type.

In [24]:
df_complete = df[df['delivery_status'] == 'Complete']
df_complete[df_complete['order_result'] == "delay"].groupby(['order_result']).agg(
    delivery_delay_on_average = ('order_difference','median')
)

,delivery_delay_on_average
order_result,
delay,8.0


According to the data, the average delivery delay is around **8 hours**.

In [25]:
df_complete[df_complete['order_result'] == "early"].groupby(['order_result']).agg(
    delivery_delay_on_average = ('order_difference','median')
)

,delivery_delay_on_average
order_result,
early,-3.0


According to the data, the average early delivery is around **3 hours**.

In [26]:
df_complete.groupby(['order_result']).agg(
    number_of_delivery = ('delivery_id','count')
)

,number_of_delivery
order_result,
delay,361
early,110
on time,25


according to the data there are 361 number delay deliveries, 110 early deliveries, 25 on time deliveries

In [27]:
df_complete.groupby(['order_result','product_type']).agg(
    number_of_delivery = ('delivery_id','count')
)

number_of_delivery
order_result product_type                    
delay        Books                         39
             Clothing                      39
             Documents                     52
             Electronics                   41
             Furniture                     45
             Groceries                     58
             Machinery                     41
             Medicine                      46
early        Books                         15
             Clothing                      11
             Documents                     13
             Electronics                   15
             Furniture                     15
             Groceries                      7
             Machinery                     18
             Medicine                      16
on time      Books                          5
             Clothing                       2
             Documents                      6
             Electronics                    2
             Furniture                      4
             Groceries                      6

According to the data, all product type have similiar number of delay delivery

In [28]:
df_complete.groupby(['order_result','delivery_way']).agg(
    number_of_delivery = ('delivery_id','count')
)

number_of_delivery
order_result delivery_way                    
delay        Air                           85
             Mixed Way                     94
             Rail                           3
             Road                          93
             Sea                           86
early        Air                           28
             Mixed Way                     29
             Road                          24
             Sea                           29
on time      Air                            6
             Mixed Way                     11
             Road                           3
             Sea                            5

According to the data, **Rail** has the smallest number of delayed deliveries compared to **Air, Mixed Way, Road, and Sea**. For early deliveries, **Air, Mixed Way, Road, and Sea** have a similar number of deliveries. For on-time deliveries, **Mixed Way** has the highest number compared to **Air, Road, and Sea**.

In [29]:
df_complete.groupby(['order_result','delivery_type']).agg(
    number_of_delivery = ('delivery_id','count')
)

number_of_delivery
order_result delivery_type                    
delay        Economy                        90
             Express                        99
             Same Day                       78
             Standard                       94
early        Economy                        35
             Express                        25
             Same Day                       23
             Standard                       27
on time      Economy                         4
             Express                         7
             Same Day                       11
             Standard                        3

According to the data, Economy, Same Day, and Standard delivery types have the highest number of delayed deliveries (90–99 delayed deliveries). Express and Same Day delivery types have the highest number of on-time deliveries. For early deliveries, the Economy delivery type has the highest number of early deliveries.

In [30]:
df_complete.groupby(['order_result','place_end']).agg(
    number_of_delivery = ('delivery_id','count')
)

number_of_delivery
order_result place_end                      
delay        Central Hub                  42
             East Center                  44
             Hub Alpha                    51
             Hub Beta                     45
             Hub Gamma                    40
             North Depot                  48
             South Depot                  49
             West Center                  42
early        Central Hub                  19
             East Center                  11
             Hub Alpha                     8
             Hub Beta                     15
             Hub Gamma                    14
             North Depot                  11
             South Depot                  20
             West Center                  12
on time      Central Hub                   1
             East Center                   3
             Hub Alpha                     5
             Hub Beta                      1
             Hub Gamma                     4
             North Depot                   3
             South Depot                   3
             West Center                   5

According to the data, all destination locations have a similar number of delayed and on-time deliveries.

In [31]:
df_complete.groupby(['delivery_type']).agg(
    rating_on_average = ('rating','mean')
)

,rating_on_average
delivery_type,
Economy,3.108527
Express,3.137405
Same Day,2.991071
Standard,2.935484


According to the data, all delivery type have a similar average of rating

In [32]:
df_complete.groupby(['delivery_type','product_type']).agg(
    rating_on_average = ('rating','mean')
)

rating_on_average
delivery_type product_type                   
Economy       Books                  3.411765
              Clothing               3.000000
              Documents              3.034483
              Electronics            3.357143
              Furniture              3.176471
              Groceries              3.153846
              Machinery              3.000000
              Medicine               2.636364
Express       Books                  2.846154
              Clothing               3.615385
              Documents              3.153846
              Electronics            3.166667
              Furniture              3.066667
              Groceries              3.071429
              Machinery              2.800000
              Medicine               3.437500
Same Day      Books                  3.250000
              Clothing               3.000000
              Documents              3.166667
              Electronics            2.818182
              Furniture              3.176471
              Groceries              3.000000
              Machinery              2.470588
              Medicine               3.055556
Standard      Books                  3.384615
              Clothing               3.294118
              Documents              2.588235
              Electronics            2.400000
              Furniture              3.200000
              Groceries              2.777778
              Machinery              3.333333
              Medicine               2.705882

According to the data, sending **medicine** using **Economy** delivery has the lowest rating compared to the other product types. For **Express** and **Same Day** delivery, sending **machinery** has the lowest rating compared to the other product types. For **Standard** delivery, sending **electronics, documents, groceries, and medicine** has the lowest rating.

In [33]:
df.head()

,delivery_id,delivery_time_start,delivery_time_finish,delivery_time_estimation_in_hours,delivery_time_result_in_hours,delivery_status,price,weight,product_type,place_start,place_end,rating,delivery_type,distance,delivery_way,refund,carrier_id,package_volume_m3,order_difference,order_result
0,DEL00001,2026-03-16 14:00,2026-03-19 02:00,60,60,Damage,368.12,216.870,Furniture,Hub Beta,Hub Beta,4,Same Day,145.6,Sea,No,CAR008,1.554,0,on time
91,DEL00092,2026-06-30 14:00,2026-06-30 20:00,6,6,Complete,325.91,280.320,Electronics,Central Hub,Hub Gamma,3,Same Day,1224.3,Air,No,CAR016,4.662,0,on time
95,DEL00096,2026-08-18 09:00,2026-08-20 08:00,47,47,Damage,276.50,317.110,Electronics,West Center,South Depot,1,Economy,896.8,Mixed Way,No,CAR004,3.885,0,on time
144,DEL00145,2026-06-24 04:00,2026-06-25 22:00,42,42,Damage,255.90,189.010,Groceries,West Center,Hub Beta,4,Express,2193.3,Road,Yes,CAR018,1.396,0,on time
157,DEL00158,2026-02-15 06:00,2026-02-17 10:00,52,52,Damage,20.00,238.855,Documents,East Center,South Depot,4,Economy,819.1,Sea,No,CAR008,1.656,0,on time


In [34]:
df.to_csv('delivery.csv')

In [35]:
df_complete.to_csv('delivert_complete.csv')